File: Src/WitsmlExplorer.Api/Jobs/Common/ComponentReferences.cs
using System;
using System.Text;

using WitsmlExplorer.Api.Models;

namespace WitsmlExplorer.Api.Jobs.Common
{
    public class ComponentReferences : IReference
    {
        public ObjectReference Parent { get; set; }
        public ComponentType ComponentType { get; set; }
        public string[] ComponentUids { get; set; } = Array.Empty<string>();

        public string Description()
        {
            StringBuilder desc = new();
            desc.Append(Parent.Description());
            desc.Append($"ComponentType: {ComponentType}; ");
            desc.Append($"ComponentUids: {string.Join(", ", ComponentUids)}; ");
            return desc.ToString();
        }

        public void Verify()
        {
            if (ComponentUids == null || ComponentUids.Length == 0)
            {
                throw new ArgumentException("A minimum of one component UID is required");
            }
            if (Parent == null)
            {
                throw new ArgumentException("Parent is required");
            }
            Parent.Verify();
        }

        public string GetObjectName()
        {
            return Parent.Name;
        }

        public string GetWellboreName()
        {
            return Parent.WellboreName;
        }

        public string GetWellName()
        {
            return Parent.WellName;
        }
    }
}

File: Src/WitsmlExplorer.Frontend/models/jobs/componentReferences.ts
import { ComponentType } from "models/componentType";
import ObjectReference from "models/jobs/objectReference";
import ObjectOnWellbore, { toObjectReference } from "models/objectOnWellbore";

export default interface ComponentReferences {
  serverUrl?: string;
  parent: ObjectReference;
  componentType: ComponentType;
  componentUids: string[];
}

export interface CopyRangeClipboard extends ComponentReferences {
  startIndex?: string;
  endIndex?: string;
}

export function createComponentReferences(
  uids: string[],
  parent: ObjectOnWellbore,
  componentType: ComponentType,
  serverUrl?: string
): ComponentReferences {
  return {
    serverUrl: serverUrl,
    parent: toObjectReference(parent),
    componentUids: uids,
    componentType
  };
}
